### PR TITLE
Some stability updates and fixes for replaceById

### PR DESCRIPTION
**Pull request by qpresley**

### CODE REVIEW

File: lib/db2.js [hunk 1]
```diff
@@ -76,6 +76,8 @@ DB2.prototype.create = function(model, data, options, callback) {
  * @param {Function} cb The callback function
  */
 DB2.prototype.update = function(model, where, data, options, cb) {
+  debug('DB2.prototype.update (enter): where=%j, data=%j, options=%j\n',
```
Comment (Amir-61) [inline lib/db2.js]: Nitpick: I do not think if `(enter)` is really needed; when you get `DB2.prototype.update` it is obvious it is already entered in this section

Comment (Amir-61) [inline lib/db2.js]: I would prefer to have this change in a separate commit with appropriate commit description; it is out of the scope of implementing `replaceById`

File: lib/db2.js [hunk 2]
```diff
@@ -88,6 +90,30 @@ DB2.prototype.update = function(model, where, data, options, cb) {
 };
 
 /**
+ * Update all instances that match the where clause with the given data
```
Comment (Amir-61) [inline lib/db2.js]: 'update all'?

File: lib/db2.js [hunk 3]
```diff
@@ -88,6 +90,30 @@ DB2.prototype.update = function(model, where, data, options, cb) {
...
+ */
+DB2.prototype.replaceById = function(model, where, data, options, cb) {
+  debug('DB2.prototype.replaceById (enter): where=%j, data=%j, options=%j\n',
+        where, data, options);
+  var self = this;
+  var stmt = self.buildReplace(model, where, data, options);
+  var id = self.idName(model);
+  var sql = 'SELECT \"' + id + '\" FROM FINAL TABLE (' + stmt.sql + ')';
+  self.execute(sql, stmt.params, options, function(err, info) {
+    if (cb) {
```
Comment (Amir-61) [inline lib/db2.js]: Why checking `cb`? AFAICT we always expect `cb` for `replaceById`

File: lib/db2.js [hunk 4]
```diff
@@ -88,6 +90,30 @@ DB2.prototype.update = function(model, where, data, options, cb) {
...
+      cb(err, {count: info.length});
```
Comment (Amir-61) [inline lib/db2.js]: Why just returning `info.length`, why not passing whole object?

File: lib/db2.js (hunk 2 above)
Comment (qpresley) [inline lib/db2.js]: While technically not wrong I will change it to "Update the instance matching the ID in the where clause".  Note that if the id field is not unique there could be more than one and loopback does not take that into account.

File: lib/db2.js [hunk 5]
```diff
@@ -88,6 +90,30 @@ DB2.prototype.update = function(model, where, data, options, cb) {
...
+    }
+  });
```
Comment (Amir-61) [inline lib/db2.js]: I would probably simplify the above block to:
```
self.execute(sql, stmt.params, options, cb);
```
Thought?

File: lib/db2.js (hunk 1 above)
Comment (qpresley) [inline lib/db2.js]: While in this example there is no exit it is nice to see enter vs. exit for other functions and sticking to a standard is helpful.

Commit descriptions are 50 characters max and I'm not sure they are useful.  This is pretty self explanatory.

File: lib/transaction.js [hunk 6]
```diff
@@ -39,15 +39,13 @@ function mixinTransaction(DB2, db2) {
...
 
-          if (sql) {
-            connection.query(sql, function(err) {
-              cb(err, connection);
-            });
-          } else {
+          connection.query(stmt, function(err) {
             cb(err, connection);
-          }
+          });
```
Comment (Amir-61) [inline lib/transaction.js]: Are these changes relevant to implementation of replaceById? Shouldn't they be in a separate commit with appropriate commit message?

File: lib/db2.js (hunk 3 above)
Comment (qpresley) [inline lib/db2.js]: Always until someone doesn't pass it in.  Removed.

File: lib/db2.js (hunk 4 above)
Comment (qpresley) [inline lib/db2.js]: There appear to be no tests that validate the return value and none of the function headers describe what data is expected to be returned.  Do we have such info and I will change this to return the appropriate info?

**Comment by Amir-61** [comment]
* Are the `nvm` failures relevant?
* TBH still I'm not sure what was wrong with general implementation of `replaceById` from `loopback-connector`; I tried that one now and it seems working and it returns `id` in the `cb` as well; am I missing something?

@superkhau @loay @0candy PTAL too

**Comment by qpresley** [comment]
@Amir-61:  Figured out where it is broken.  I'm working on a DB2 iSeries connector which the UPDATE statement in buildReplace will not work for as it needs an isolation level appended otherwise it will hang.  I'll delete this branch for now and you can consider the replaceById work closed.
